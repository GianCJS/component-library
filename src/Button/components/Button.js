import React from 'react';
import PropTypes from 'prop-types';

function Button(props) {
    const {
        primary, secondary, transparent, type,
        children, content, className, onClick,
        disabled, fluid, qaData, loading,
    } = props;

    let { size } = props;
    const allowedSizes = ['small', 'normal', 'large'];
    size = allowedSizes.some(validSize => validSize === size) ? `size-${size}` : '';

    const buttonClassName = `button
        ${primary ? ' primary-button' : ''}
        ${secondary ? ' secondary-button' : ''}
        ${transparent ? ' transparent-button' : ''}
        ${fluid ? 'fluid-button' : ''}
        ${loading ? 'is-loading-button' : ''}
        ${size}
        ${className}`;

    return (
        <button // eslint-disable-line react/button-has-type
            className={buttonClassName}
            onClick={onClick}
            type={type}
            disabled={disabled}
            {...qaData}
        >
            <span className="button-content">
                { content || children }
            </span>
            { loading && <span className="button-loader" /> }
        </button>
    );
}

Button.defaultProps = {
    content     : '',
    children    : '',
    primary     : false,
    secondary   : false,
    transparent : false,
    disabled    : false,
    className   : '',
    onClick     : null,
    type        : 'button',
    size        : 'normal',
    fluid       : false,
    loading     : false,
    qaData      : {},
};

Button.propTypes = {
    /**
    Button content
    */
    content     : PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
    /**
    Button content
    */
    children    : PropTypes.oneOfType([PropTypes.node, PropTypes.string]),
    /**
    Button with max emphasis
    */
    primary     : PropTypes.bool,
    /**
    Button with medium emphasis
    */
    secondary   : PropTypes.bool,
    /**
    Button with lower emphasis
    */
    transparent : PropTypes.bool,
    /**
    Button with disabled state
    */
    disabled    : PropTypes.bool,
    /**
    Button custom className
    */
    className   : PropTypes.string,
    /**
    onClick function
    */
    onClick     : PropTypes.func,
    /**
    Button html type
    */
    type        : PropTypes.oneOf(['button', 'submit', 'reset']),
    /**
    Button size
    */
    size        : PropTypes.string,
    /**
    Full width button
    */
    fluid       : PropTypes.bool,
    /**
    Button with loading state
    */
    loading     : PropTypes.bool,
    /**
    Give QA params
    */
    qaData      : PropTypes.shape({}),
};

export default Button;
