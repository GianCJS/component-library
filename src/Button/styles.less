@button-default-background: #eaeaea;

@button-primary-background: @color-main;
@button-primary-color: #ffffff;

@button-secondary-color: @button-primary-background;
@button-secondary-hover-background: @button-primary-background;
@button-secondary-hover-color: @button-primary-color;

@button-padding-x: @space-normal;
@button-padding-y: @space-small;

/* Main */

.button {
    box-sizing: border-box;
    border-width: 1px;
    border-style: solid;
    border-color: rgba(0, 0, 0, .1);
    border-radius: @base-border-radius;
    background: @button-default-background;
    color: @color-text-main;
    cursor: pointer;
    transition: all .2s;
    display: inline-flex;
    justify-content: center;
    align-items: center;
    -webkit-appearance: none !important;
    -moz-appearance: none !important;
    appearance: none !important;
    outline: 0;
    font-weight: bold;
    transition: all .2s ease-in-out;
    position: relative;
    padding: @button-padding-y @button-padding-x;

    &:hover {
        background: darken(@button-default-background, 5%);
        transition: all .2s ease-in-out;
    }

    &:focus,
    &:active {
        background: darken(@button-default-background, 10%);  
        transition: all .2s ease-in-out;      
    }

    &.is-loading-button {

        &:hover,
        &:active,
        &:focus {
            background: @button-default-background;
        }
        
    }

    &[disabled],
    &:disabled {
        opacity: .4;
        cursor: default;
    }

    /* Fluid */

    &.fluid-button {
        width: 100%;
    }

    /* Size */

    &.size-small {
        font-size: @font-size-small;
        padding: (@button-padding-y * .75) @button-padding-x;
    }    

    &.size-normal {
        font-size: @font-size-normal;        
    }

    &.size-large {
        font-size: @font-size-medium;
        padding: (@button-padding-y * 1.5) (@button-padding-x * 1.25);
    }

    /* Primary */   
    
    &.primary-button {
        background: @button-primary-background;
        color: @button-primary-color;
        border-color: transparent;
        text-transform: uppercase; 
        
        &:not([disabled]),
        &:not(:disabled) {
        
            &:hover {
                background: darken(@button-primary-background, 5%);            
            }
            
            &:focus,
            &:active {
                background: darken(@button-primary-background, 10%);            
            }

        }

        &.is-loading-button {

            &:hover,
            &:active,
            &:focus {
                background: @button-primary-background;
            }
            
        }

        .button-loader {
            border-color: fadeout(@button-primary-color, 70%);
        }
    
    }

    /* Secondary */
    
    &.secondary-button {
        border-color: @button-secondary-color;
        color: @button-secondary-color;
        background: transparent;        

        &:not([disabled]),
        &:not(:disabled) {
        
            &:hover {
                color: @button-secondary-hover-color;
                background: @button-secondary-hover-background;
                border-color: transparent;

                .button-loader {
                    border-color: fadeout(@button-secondary-hover-color, 70%);
                }

            }
            
            &:focus,
            &:active {
                color: @button-secondary-hover-color;
                background: darken(@button-secondary-hover-background, 10%);

                .button-loader {
                    border-color: fadeout(@button-secondary-hover-color, 70%);
                }

            }

        }

        &.is-loading-button {

            &:hover,
            &:active,
            &:focus {
                border-color: @button-secondary-color;
                color: @button-secondary-color;
                background: transparent;

                .button-loader {
                    border-color: fadeout(@button-secondary-color, 70%);
                }

            }
            
        }

        .button-loader {
            border-color: fadeout(@button-secondary-color, 70%);
        }
    
    }

    /* Transparent */
    
    &.transparent-button {
        border-color: transparent;
        color: @button-primary-background;
        background: transparent;
        text-transform: uppercase;        

        &:not([disabled]),
        &:not(:disabled) {
        
            &:hover {
                color: darken(@button-primary-background, 5%);
                background: fade(@button-primary-background, 7%);

                .button-loader {
                    border-color: fadeout(@button-primary-background, 70%);
                }
                
            }
            
            &:focus,
            &:active {
                color: darken(@button-primary-background, 10%);
                background: fade(@button-primary-background, 12%);

                .button-loader {
                    border-color: fadeout(@button-primary-background, 70%);
                }

            }

        }

        &.is-loading-button {

            &:hover,
            &:active,
            &:focus {
                border-color: transparent;
                color: @button-primary-background;
                background: transparent;
            }

        }

        .button-loader {
            border-color: fadeout(@button-primary-background, 70%);
        }
    
    }

    /* Icons */
    
    i[class^="bf-icon-"] {
        font-weight: bold;
        margin-right: @space-small;
    }

}

/* Loading state */

.button-loader {
    position: absolute;
    transform: translate(-50%, -50%);
    content: "";
    height: @font-size-normal;
    width: @font-size-normal;
    border-radius: 50%;
    border-width: 3px;
    border-style: solid;
    border-color: lightgray;
    border-top-color: currentColor !important;
    animation: loading-spinner .7s linear infinite;
    visibility: hidden; 
}

@keyframes loading-spinner {

    from {
        transform: rotate(0deg);
    }

    to {
        transform: rotate(360deg);
    }

}

.button.is-loading-button {
    cursor: default;

    .button-content {
        visibility: hidden;
    }

    .button-loader {
        visibility: visible;
    }

}
