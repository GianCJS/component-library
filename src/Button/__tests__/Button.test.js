import React from 'react';
import { shallow } from 'enzyme';
import Button from '../index';

describe('Button Unit Test', () => {
    let props;
    let shallowedButton;

    const getButton = () => {
        if (!shallowedButton) {
            shallowedButton = shallow(<Button {...props} />);
        }
        return shallowedButton;
    };

    beforeEach(() => {
        props = {};
        shallowedButton = undefined;
    });

    test('Component should render without crashing', () => {
        expect(getButton()).toBeDefined();
    });

    test('Component should render correctly', () => {
        expect(getButton()).toMatchSnapshot();
    });
});
