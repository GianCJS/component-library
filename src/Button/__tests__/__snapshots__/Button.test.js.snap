// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Unit Test Component should render correctly 1`] = `
<button
  className="button
        
        
        
        
        
        size-normal
        "
  disabled={false}
  onClick={null}
  type="button"
>
  <span
    className="button-content"
  />
</button>
`;
